### PR TITLE
[String] Bug fix for opaque slicing in comparison

### PR DESCRIPTION
**Pull request by milseman**
Starting a PR with a minor re-organization attempting to reduce bugs and make `_asOpaque` only appear in non-inlineable code.

### CODE REVIEW

**Comment by milseman** [comment]
@lancep can you take this over and add test case that fails without this? This fixes the issue where we ignore the trailing end of the slice for pathological or complex comparisons.

**Comment by milseman** [comment]
(this reorganization specifically is to unblock some of my small string work, but might as well fix bugs too!)

**Comment by milseman** [comment]
Just to see if it's still good:

@swift-ci please test

**Comment by milseman** [comment]
(Not expecting much here)

@swift-ci please smoke benchmark

**Comment by milseman** [comment]
Linux failure might be the expected failure until after https://github.com/apple/swift-corelibs-foundation/pull/1441#issuecomment-366554616 is merged
```
Test Case 'TestNSString.test_PrefixSuffix' started at 2018-02-18 22:55:45.216
TestFoundation/TestNSString.swift:1412: error: TestNSString.test_PrefixSuffix : XCTAssertTrue failed - Unexpected success: 1241
```

File: stdlib/public/core/StringComparison.swift [hunk 1]
```diff
@@ -16,25 +16,49 @@ import SwiftShims
...
+@_versioned // @opaque
+internal
 func _compareUnicode(
   _ lhs: _StringGuts._RawBitPattern, _ rhs: _StringGuts._RawBitPattern
 ) -> Int {
   let left = _StringGuts(rawBits: lhs)
   let right = _StringGuts(rawBits: rhs)
-  return left._compare(right)
+
+   if _slowPath(!left._isContiguous || !right._isContiguous) {
```
Comment (milseman) [inline stdlib/public/core/StringComparison.swift]: Derp, this might be over-indented by one space

**Comment by lancep** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please test

**Comment by lancep** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please test

**Pull request merged**